linera-core/src/client/validator_manager/request.rs

```diff
@@ -0,0 +1,235 @@
+    Certificate(Box<ConfirmedBlockCertificate>),
+}
+
+impl From<Option<Blob>> for RequestResult {
+    fn from(blob: Option<Blob>) -> Self {
+        RequestResult::Blob(blob)
+    }
+}
+
+impl From<Vec<ConfirmedBlockCertificate>> for RequestResult {
+    fn from(certs: Vec<ConfirmedBlockCertificate>) -> Self {
+        RequestResult::Certificates(certs)
+    }
+}
+
+impl From<BlobContent> for RequestResult {
+    fn from(content: BlobContent) -> Self {
+        RequestResult::BlobContent(content)
+    }
+}
+
+impl From<ConfirmedBlockCertificate> for RequestResult {
+    fn from(cert: ConfirmedBlockCertificate) -> Self {
+        RequestResult::Certificate(Box::new(cert))
+    }
+}
+
+impl From<RequestResult> for Option<Blob> {
+    fn from(result: RequestResult) -> Self {
+        match result {
+            RequestResult::Blob(blob) => blob,
+            _ => panic!("Cannot convert RequestResult to Option<Blob>"),
+        }
+    }
+}
+
+impl From<RequestResult> for Vec<ConfirmedBlockCertificate> {
+    fn from(result: RequestResult) -> Self {
+        match result {
+            RequestResult::Certificates(certs) => certs,
+            _ => panic!("Cannot convert RequestResult to Vec<ConfirmedBlockCertificate>"),
+        }
+    }
+}
+
+impl From<RequestResult> for BlobContent {
+    fn from(result: RequestResult) -> Self {
+        match result {
+            RequestResult::BlobContent(content) => content,
+            _ => panic!("Cannot convert RequestResult to BlobContent"),
+        }
+    }
+}
+
+impl From<RequestResult> for ConfirmedBlockCertificate {
+    fn from(result: RequestResult) -> Self {
+        match result {
+            RequestResult::Certificate(cert) => *cert,
+            _ => panic!("Cannot convert RequestResult to ConfirmedBlockCertificate"),
+        }
+    }
+}
+
+impl SubsumingKey<RequestResult> for super::request::RequestKey {
+    fn subsumes(&self, other: &Self) -> bool {
+        // Different chains can't subsume each other
+        if self.chain_id() != other.chain_id() {
+            return false;
+        }
+
+        let heights1 = match self.height_range() {
+            Some(range) => range,
+            None => return self == other, // Non-certificate requests must match exactly
+        };
+        let heights2 = match other.height_range() {
+            Some(range) => range,
+            None => return false, // Can't subsume different variant types
+        };
+
+        // Check if all heights in other are contained in self
+        heights2.iter().all(|h| heights1.contains(h))
+    }
+
+    fn try_extract_result(&self, from: &Self, result: &RequestResult) -> Option<RequestResult> {
+        // Only certificate results can be extracted
+        let certificates = match result {
+            RequestResult::Certificates(certs) => certs,
+            _ => return None,
+        };
+
+        if self.chain_id().is_none() || from.chain_id().is_none() {
+            return None;
+        }
```

Comment:
Instead of all this, shouldn't we just `if !self.subsumes(result) { return None; }`?

Otherwise it isn't guaranteed that the `filtered` results below are complete.